# A simple application to get all the anagrams from a given list

### a simple list which contains some anagrams

In [39]:
words=["cat", "dog", "tac", "god", "act"]

In [30]:
words=words+['guff','dgf','dffu']

In [31]:
sorted(list(set(words)))

['act', 'cat', 'dffu', 'dgf', 'dog', 'god', 'guff', 'tac']

In [9]:
def sort(word):
    return "".join(sorted(word))

## fucntional approach

In [16]:
def anagram(myword):
    return [word for word in words if sign(word)==sign(myword)]

In [17]:
anagram("cta")

['cat', 'tac', 'act']

In [18]:
%timeit anagram("cta")

10000 loops, best of 3: 26.1 µs per loop


# efficent approach

In [19]:
from collections import defaultdict

In [32]:
mydict=defaultdict(list)

In [33]:
for word in words:
    mydict[sign(word)].append(word)

In [34]:
mydict


defaultdict(list,
            {'act': ['cat', 'tac', 'act'],
             'dffu': ['dffu'],
             'dfg': ['dgf'],
             'dgo': ['dog', 'god'],
             'ffgu': ['guff']})

In [35]:
words


['cat', 'dog', 'tac', 'god', 'act', 'guff', 'dgf', 'dffu']

In [36]:
def ana(myword):
    return mydict[sign(myword)]


In [37]:
ana('tac')

['cat', 'tac', 'act']

In [38]:
ana('dog')

['dog', 'god']

In [40]:
%timeit ana('dog')

The slowest run took 9.72 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 1.13 µs per loop


### Overall time speed reduced from 26.1 micro seconds to 1.13 micro seconds...woahhhhhhhhhhh!!!